client/consensus/grandpa/authorities.go

```diff
@@ -0,0 +1,814 @@
+
+		status.newSetBlock = &hashNumber[H, N]{
+			hash:   finalizedHash,
+			number: finalizedNumber,
+		}
+	}
+
+	return status, nil
+}
+
+// EnactsStandardChange Check whether the given finalized block number enacts any standard
+// authority set hashNumber (without triggering it), ensuring that if there are
+// multiple changes in the same branch, finalizing this block won't
+// finalize past multiple transitions (i.e. transitions must be finalized
+// in-order). Returns *true if the block being finalized enacts a
+// hashNumber that can be immediately applied, *false if the block being
+// finalized enacts a hashNumber but it cannot be applied yet since there are
+// other dependent changes, and nil if no hashNumber is enacted. The given
+// function `is_descendent_of` should return `true` if the second hash
+// (target) is a descendent of the first hash (base).
+func (authSet *AuthoritySet[H, N]) EnactsStandardChange(
+	finalizedHash H, finalizedNumber N, isDescendentOf IsDescendentOf[H]) (*bool, error) {
+	applied, err := authSet.pendingStandardChanges.FinalizesAnyWithDescendentIf(&finalizedHash, finalizedNumber, isDescendentOf, func(change *PendingChange[H, N]) bool {
+		return change.EffectiveNumber() == finalizedNumber
+	})
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", errForkTree, err)
+	}
+	return applied, nil
+}
+
+// PendingChange A pending hashNumber to the authority set.
+//
+// This will be applied when the announcing block is at some depth within
+// the finalized or unfinalized chain.
+type PendingChange[H comparable, N constraints.Unsigned] struct {
+	// The new authorities and weights to apply.
+	nextAuthorities []Authority
+	// How deep in the chain the announcing block must be
+	// before the hashNumber is applied.
+	delay N
+	// The announcing block's height.
+	canonHeight N
+	// The announcing block's hash.
+	canonHash H
+	// The delay kind.
+	delayKind delayKind
+}
+
+// EffectiveNumber Returns the effective number this hashNumber will be applied at.
+func (pc *PendingChange[H, N]) EffectiveNumber() N {
+	return pc.canonHeight + pc.delay
+}
+
+// AuthoritySetChanges Tracks historical authority set changes. We store the block numbers for the last block
+// of each authority set, once they have been finalized. These blocks are guaranteed to
+// have a justification unless they were triggered by a forced hashNumber.
+type AuthoritySetChanges[N constraints.Unsigned] []setIDNumber[N]
+
+// append an setIDNumber to AuthoritySetChanges
+func (asc *AuthoritySetChanges[N]) append(setId uint64, blockNumber N) {
+	*asc = append(*asc, setIDNumber[N]{
+		setId:       setId,
+		blockNumber: blockNumber,
+	})
+}
+
+// Three states that can be returned: Latest, Set (tuple), Unknown
+// Latest => bool
+// Set => &AuthorityChange
+// Unknown => nil
+// TODO for reviewers, this can be a VDT but I'm not sure its needed
+func (asc *AuthoritySetChanges[N]) getSetID(blockNumber N) (latest bool, set *setIDNumber[N], err error) {
```

Comment:
I think we should recreate the enum in this case, even though the `AuthoritySetChangeId` type in the rust code isn't actually exported outside of the crate, I think it makes it easier to cross reference.  I also think dealing with more than 1 non-error return value is kinda odd.